### ISSUE TITLE
Spike: how much customization can we do of the Autofill capture UI?

### ISSUE THREAD

**Issue by jhugman** [mozilla-lockwise/lockwise-android#1105]
### Vision statement / What / Requirements

Issue #1081 asks for some quite extensive user interaction in the Capture Credential From Autofill:

 * show the username and obscured password
 * toggle visibility of obscured password
 * edit username and password before saving.

This issue is the research to see how much of this is possible.

**Comment by eliserichards** [comment]
What is the priority of this? @sandysage @changecourse

**Comment by sandysage** [comment]
I'm marking as a P2. I'd like to tee this up for the in-between time sprint here at the end of the year. Let's see where we can fit this in after the next release.

**Comment by jhugman** [comment]
1105 — Research spike on how to do a fenix like "save login" door hanger.
https://github.com/mozilla-lockwise/lockwise-android/issues/1105

## TL;DR: 

We can likely show the username and password that the user is about to save. Other user-controls will not be possible.

## Detail

This is done in android-components as part of the [`PromptsFeature`][pf]. The XML is at [`layout/mozac_feature_prompt_login_prompt.xml`][xml].

There are two `EditText` fields, allowing for the user to edit the username and password, an `ImageViewButton` which allows the user to toggle the visibility of the password, and two `Button`s to "Save" or "Don't Save". There are two `TextView`s which show generic text and the hostname of the site being saved.

Saving a credential with the Android Autofill Manager, requires `LockboxAutofillService` adds a `SaveInfo` to the `FillRequest` _before_ any logins has been offered to the user. 

Any [UI customization to the save dialog][ui] is done here in the `SaveInfo`, via a [`CustomDescription`][cd] and [`RemoteViews`][rv].

However, `RemoteViews` are fairly limited in their support for user display and interaction:

```
RemoteViews is limited to support for the following layouts:

    AdapterViewFlipper
    FrameLayout
    GridLayout
    GridView
    LinearLayout
    ListView
    RelativeLayout
    StackView
    ViewFlipper

And the following widgets:

    AnalogClock
    Button
    Chronometer
    ImageButton
    ImageView
    ProgressBar
    TextClock
    TextView

Descendants of these classes are not supported.
```

This would make editing the username and password almost impossible.

Additionally, it is not easy/possible to see the `RemoteViews` of the dialog once it is displayed. This makes it difficult to change the UI after the dialog is displayed (e.g. tapping on a visibility toggle).

[xml]: https://github.com/mozilla-mobile/android-components/blob/b92c96fad88af738e4df25b921e9b8542f6f1d56/components/feature/prompts/src/main/res/layout/mozac_feature_prompt_login_prompt.xml
[cd]: https://developer.android.com/reference/android/service/autofill/CustomDescription.html
[pf]: https://github.com/mozilla-mobile/android-components/blob/795c46cf789abbe315a2df538d03cf83a459112f/components/feature/prompts/README.md
[rv]: https://developer.android.com/reference/android/widget/RemoteViews.html
[ui]: https://developer.android.com/reference/android/service/autofill/SaveInfo.html#customizing-the-autofill-save-ui

**Comment by sandysage** [comment]
@changecourse Let's discuss these findings and determine how/if this affects our approach for the Fenix integration. Closing this issue as complete.
